dayforce_client/client.py

```diff
@@ -105,17 +110,83 @@ class DayforceSFTP(object):
     disable_host_key_checking: bool = attr.ib(default=False)
     cnopts: pysftp.CnOpts = attr.ib(init=False)
 
+    _sftp: Optional[pysftp.Connection] = attr.ib(default=None)
+    _sftp_live: bool = attr.ib(default=False)
+
     def __attrs_post_init__(self):
         self.cnopts = pysftp.CnOpts()
         if self.disable_host_key_checking is True:
             self.cnopts.hostkeys = None
         elif self.host_key is not None:
             key_b = self.host_key.encode()
             key = paramiko.RSAKey(data=base64.decodebytes(key_b))
-            self.cnopts.hostkeys.add(self.hostname, 'ssh-rsa', key)
+            self.cnopts.hostkeys.add(self.hostname, "ssh-rsa", key)
+        else:
+            raise RuntimeError("disable_host_key_checking or host_key must be set")
+
+    def _connect(self):
+        """Establish the SFTP connection."""
+        if not self._sftp_live:
+            self._sftp = pysftp.Connection(
+                host=self.hostname,
+                username=self.username,
+                password=self.password,
+                port=self.port,
+                cnopts=self.cnopts,
+            )
+            self._sftp_live = True
+
+    def _disconnect(self):
+        """Close the SFTP connection."""
+        if self._sftp_live:
+            self._sftp.close()
+            self._sftp_live = False
+
+    @contextmanager
+    def connect(self):
+        self._connect()
+        try:
+            yield self
+        except Exception as e:
+            self._disconnect()
+            raise e
+
+    """Upload a batch import file and return a token for status checking"""
+
+    def put_import(self, filename: str, type: str) -> str:
+        self._connect()
+        remotepath = f"/Import/{type}/{filename}"
+        if os.path.getsize(filename) > 100 * 1e6:
+            raise RuntimeError(f"{filename} exceeds the 100MB batch size limit")
+        self._sftp.put(filename, remotepath=remotepath)
+        self._sftp.rename(remotepath, f"{remotepath}.ready")
+        return remotepath
+
+    def raise_for_import_status(self, token: str):
+        self._connect()
+        filename = os.path.basename(token)
+        dirname = os.path.dirname(token)
+        if self._sftp.exists(f"{dirname}/archive/{filename}.done"):
+            return
+        elif self._sftp.exists(f"{dirname}/error/{filename}.error"):
+            raise ImportError()
         else:
-            raise RuntimeError('disable_host_key_checking or host_key must be set')
+            raise ImportPending()
+
+
+class Error(Exception):
```

Comment:
@bobzoller what are your thoughts on renaming this something to a little more explicit? Maybe `DayforceError` instead of `Error`?

```suggestion
class DayforceError(Exception):
```

Reply:
updated